File: go/internal/twiml/verb_dial.go
package twiml

import (
	"context"
	"time"
)

// dialHandler dispatches <Dial> mid-call. Terminal verb: Twilio's model
// specifies that <Dial> stops the verb chain; subsequent verbs in the document
// are unreachable. dialHandler returns after the dial completes (success or
// failure) by calling t.Terminate with the appropriate reason.
//
// Privacy Gate ordering:
//  1. t.PrepareDial     — closes WS stream FIRST, then allocates callee leg + RTP port
//  2. t.PerformDial     — outbound INVITE → dialog → BYE; fires action callback
//  3. t.Terminate       — stamps termReason on the caller leg; Twilio REST clients
//     see the final status via Status() on the next GET /Calls/{Sid}.json
//
// Handle release is deferred immediately after PrepareDial succeeds so RTP
// ports are freed regardless of the PerformDial outcome (TWIML-04 / threat
// model: RTP port leak when PerformDial fails).
func dialHandler(ctx context.Context, dial *Dial, t DialTarget) error {
	t.Log().Info().Msg("twiml: dialHandler invoked — resolving <Dial> target")

	target, ambiguous := dial.ResolveDialTarget()
	if ambiguous {
		t.Log().Warn().Msg("twiml: <Dial> has both bare-text and <Number> child — <Number> wins (Twilio precedence)")
	}
	if target == "" {
		t.Log().Warn().Msg("twiml: <Dial> has no target — skipping")
		return nil
	}

	t.Log().Info().
		Str("target", target).
		Bool("hangup_on_star", dial.HangupOnStar).
		Str("action", dial.Action).
		Msg("twiml: <Dial> target resolved — invoking PrepareDial")
	if dial.HasSip || dial.HasClient || dial.HasConference || dial.HasQueue {
		t.Log().Warn().Msg("twiml: <Dial> has unsupported child (Sip/Client/Conference/Queue) — only <Number> + bare-text are supported")
		// Continue with the resolved target if any; handler falls through to dial it.
		// If the target resolved from NumberText/Number is empty, the early-return above applies.
	}

	// Resolve per-<Number> overrides over parent <Dial>. Twilio's
	// documented per-leg precedence: if a <Number statusCallback*>
	// attribute is non-empty, it wins. Empty values fall back to the
	// parent <Dial> attribute.
	statusCallback := dial.StatusCallback
	statusCallbackMethod := dial.StatusCallbackMethod
	statusCallbackEvents := dial.StatusCallbackEvents
	if dial.Number != nil {
		if dial.Number.StatusCallback != "" {
			statusCallback = dial.Number.StatusCallback
		}
		if dial.Number.StatusCallbackMethod != "" {
			statusCallbackMethod = dial.Number.StatusCallbackMethod
		}
		if len(dial.Number.StatusCallbackEvents) > 0 {
			statusCallbackEvents = dial.Number.StatusCallbackEvents
		}
	}

	opts := DialOpts{
		CallerID:             dial.CallerID,
		Timeout:              time.Duration(intPtrOr(dial.Timeout, 30)) * time.Second,
		TimeLimit:            time.Duration(intPtrOr(dial.TimeLimit, 14400)) * time.Second,
		HangupOnStar:         dial.HangupOnStar,
		Action:               dial.Action,
		Method:               dial.Method,
		StatusCallback:       statusCallback,
		StatusCallbackMethod: statusCallbackMethod,
		StatusCallbackEvents: statusCallbackEvents,
	}

	handle, err := t.PrepareDial(opts)
	if err != nil {
		t.Log().Error().Err(err).Msg("twiml: PrepareDial failed — terminating caller with reason=failed")
		return t.Terminate("failed")
	}
	defer handle.Release()

	result, err := t.PerformDial(ctx, target, opts, handle)
	if err != nil {
		t.Log().Error().Err(err).Msg("twiml: PerformDial failed — terminating caller")
		reason := twilioReasonFromDialResult(result, err)
		return t.Terminate(reason)
	}

	return t.Terminate(twilioReasonFromDialResult(result, nil))
}

// intPtrOr returns *p if non-nil, else def.
func intPtrOr(p *int, def int) int {
	if p == nil {
		return def
	}
	return *p
}

// twilioReasonFromDialResult maps DialResult.Status (or error fallback) to a
// bridge.CallSession termReason string that twilioStatusFromTermReason then
// maps to the Twilio CallStatus wire-format enum.
//
//	"answered"     → "completed"   (talk time elapsed; callee BYE'd)
//	"busy"         → "busy"        (callee 486/600)
//	"no-answer"    → "no-answer"   (ring timeout / 408/480)
//	"canceled"     → "canceled"    (caller hung up during ring)
//	"hangup-star"  → "completed"   (caller pressed '*')
//	nil result     → "failed"      (error before dial attempted)
//	default        → "failed"      (unrecognized status)
func twilioReasonFromDialResult(r *DialResult, err error) string {
	if r == nil {
		return "failed"
	}
	switch r.Status {
	case "answered":
		return "completed"
	case "busy":
		return "busy"
	case "no-answer":
		return "no-answer"
	case "canceled":
		return "canceled"
	case "hangup-star":
		return "completed"
	default:
		return "failed"
	}
}

File: node/src/twiml/verbDial.ts
/**
 * <Dial> verb handler. Port of go/internal/twiml/verb_dial.go.
 *
 * Privacy Gate ordering:
 *   1. prepareDial — closes WS stream FIRST, then allocates callee leg + RTP port
 *   2. performDial — outbound INVITE → dialog → BYE; fires action callback
 *   3. terminate   — stamps the terminal reason on the caller leg
 *
 * The handle is released regardless of the performDial outcome so RTP ports are
 * never leaked.
 */

import type { DialOpts, DialResult, DialTarget } from './dispatch.js';
import type { Dial } from './verbs.js';
import { resolveDialTarget } from './verbs.js';

const DEFAULT_TIMEOUT_MS = 30_000; // Twilio default ring timeout: 30s
const DEFAULT_TIME_LIMIT_MS = 14_400_000; // Twilio default talk time: 4h
const DEFAULT_METHOD = 'POST';

/**
 * dialHandler dispatches <Dial> mid-call. Terminal verb: resolves the target,
 * builds DialOpts (applying defaults + per-<Number> overrides), runs the
 * prepare/perform/terminate sequence, and always releases the handle.
 */
export async function dialHandler(dial: Dial, target: DialTarget): Promise<void> {
  target.log().info('twiml: dialHandler invoked — resolving <Dial> target');

  const { target: dialed, ambiguous } = resolveDialTarget(dial);
  if (ambiguous) {
    target
      .log()
      .warn('twiml: <Dial> has both bare-text and <Number> child — <Number> wins (Twilio precedence)');
  }
  if (dialed === '') {
    target.log().warn('twiml: <Dial> has no target — skipping');
    return;
  }

  target
    .log()
    .info(
      { target: dialed, hangupOnStar: dial.hangupOnStar, action: dial.action },
      'twiml: <Dial> target resolved — invoking prepareDial',
    );
  if (dial.hasSip || dial.hasClient || dial.hasConference || dial.hasQueue) {
    target
      .log()
      .warn(
        'twiml: <Dial> has unsupported child (Sip/Client/Conference/Queue) — only <Number> + bare-text are supported',
      );
  }

  // Per-<Number> overrides win over the parent <Dial> values (Twilio per-leg
  // precedence): a non-empty <Number> attribute beats the parent.
  let statusCallback = dial.statusCallback;
  let statusCallbackMethod = dial.statusCallbackMethod;
  let statusCallbackEvents = dial.statusCallbackEvents;
  if (dial.number !== undefined) {
    if (dial.number.statusCallback !== undefined && dial.number.statusCallback !== '') {
      statusCallback = dial.number.statusCallback;
    }
    if (
      dial.number.statusCallbackMethod !== undefined &&
      dial.number.statusCallbackMethod !== ''
    ) {
      statusCallbackMethod = dial.number.statusCallbackMethod;
    }
    if (
      dial.number.statusCallbackEvents !== undefined &&
      dial.number.statusCallbackEvents.length > 0
    ) {
      statusCallbackEvents = dial.number.statusCallbackEvents;
    }
  }

  const opts: DialOpts = {
    callerId: dial.callerId,
    timeoutMs: dial.timeout !== undefined ? dial.timeout * 1000 : DEFAULT_TIMEOUT_MS,
    timeLimitMs: dial.timeLimit !== undefined ? dial.timeLimit * 1000 : DEFAULT_TIME_LIMIT_MS,
    hangupOnStar: dial.hangupOnStar,
    action: dial.action,
    method: dial.method !== undefined && dial.method !== '' ? dial.method : DEFAULT_METHOD,
    statusCallback,
    statusCallbackMethod:
      statusCallbackMethod !== undefined && statusCallbackMethod !== ''
        ? statusCallbackMethod
        : DEFAULT_METHOD,
    statusCallbackEvents,
  };

  let handle;
  try {
    handle = await target.prepareDial(opts);
  } catch (err) {
    target.log().error({ err }, 'twiml: prepareDial failed — terminating caller with reason=failed');
    await target.terminate('failed');
    return;
  }

  try {
    let result: DialResult | undefined;
    try {
      result = await target.performDial(dialed, opts, handle);
    } catch (err) {
      target.log().error({ err }, 'twiml: performDial failed — terminating caller');
      await target.terminate(twilioReasonFromDialResult(undefined));
      return;
    }
    await target.terminate(twilioReasonFromDialResult(result));
  } finally {
    handle.release();
  }
}

/**
 * twilioReasonFromDialResult maps DialResult.status to the termination reason:
 *   "answered"    → "completed"
 *   "busy"        → "busy"
 *   "no-answer"   → "no-answer"
 *   "canceled"    → "canceled"
 *   "hangup-star" → "completed"
 *   undefined     → "failed"
 *   default       → "failed"
 */
function twilioReasonFromDialResult(r: DialResult | undefined): string {
  if (r === undefined) return 'failed';
  switch (r.status) {
    case 'answered':
      return 'completed';
    case 'busy':
      return 'busy';
    case 'no-answer':
      return 'no-answer';
    case 'canceled':
      return 'canceled';
    case 'hangup-star':
      return 'completed';
    default:
      return 'failed';
  }
}
